### DISCUSSION TITLE
Provide `Component.getServerSideProps` in `getInitialProps`

### DISCUSSION THREAD

**Discussion by KATT** [vercel/next.js#28183] [Ideas]
### Describe the feature you'd like to request

I'm a library author of a lib that does server-side rendering using `getInitialProps` and would like to access pages' `getServerSideProps`

The way it works right now in my library:

1. Wrap `_app` default export with a HOC
2. Makes a prepass of the whole app tree to fetch underlying queries
3. Puts those queries in a cache once they're done
4. Renders page with a hydrated cache

When using `getInitialProps` in my `_app.tsx` I'd like to be able to resolve the page component's `getServerSideProps` in order to do combine a `getInitialProps`-call which does a prepass and renders the tree but that could also get the data that the specific page component has requested.

### Related issues that this would resolve:

- https://github.com/trpc/trpc/issues/596 (my project)
- https://github.com/FormidableLabs/urql/discussions/1091

### Describe the solution you'd like

It would be nice if 

```tsx
type AppContextType<R extends NextRouter = NextRouter> = {
    Component: NextComponentType<NextPageContext>;
    AppTree: AppTreeType;
    ctx: NextPageContext;
    router: R;
};
```

Was extended to something that contained the component's `GetServerSidePropsContext`-function, example:

```tsx
type AppContextType<R extends NextRouter = NextRouter, Q extends ParsedUrlQuery = ParsedUrlQuery> = {
    Component: NextComponentType<NextPageContext>;
    ComponentGetServerSideProps: GetServerSidePropsContext<Q>;
    AppTree: AppTreeType;
    ctx: NextPageContext;
    router: R;
};
```

### Describe alternatives you've considered

There might be a way of explicitly loading a component's `getServerSideProps` function in my library's root HOC that I add to the `_app` - maybe using something like [`loadComponents`](https://github.com/vercel/next.js/blob/4dad4fb7bf6ece11af8d606129d651dec540bf66/packages/next/server/load-components.ts#L56). Is there perhaps a way of doing that?

---

### Extra notes

This all currently works with stale-while-revalidate edge-caching on Vercel - to see a reference project have a look at https://typescript.careers (source is in the footer)

**Comment by tahanawab4848** [comment]
Hi @KATT,

This is a great point and a well-known friction point for libraries handling SSR pre-passes. 

While extending `AppContextType` to include the page's `getServerSideProps` would be incredibly convenient for `_app` HOCs, Next.js intentionally isolates `getServerSideProps` to maintain strict compilation boundaries. 

### 🧱 The Core Blocker: Tree-Shaking
`getServerSideProps` is strictly server-side code. During the `next build` process, Next.js forcefully strips it from the client-side JavaScript bundle to ensure sensitive logic (like direct database queries or private API keys) never leaks to the browser. 

Because `_app` and `getInitialProps` execute on **both** the server and the client (during client-side routing), passing `getServerSideProps` into the `AppContext` risks bundling that server-only code into the client, compromising Next.js's security guarantees.

### 🛠️ Standard Ecosystem Workaround
Due to this hard boundary, the standard pattern relies on **Server-Side Helpers** invoked explicitly inside the page's `getServerSideProps`.

```typescript
// pages/index.tsx
export const getServerSideProps: GetServerSideProps = async (context) => {
  // 1. Initialize the library's server-side helper
  const helpers = createServerSideHelpers({ router: appRouter, ctx: {} });

  // 2. Perform the pre-pass / fetch the data
  await helpers.myQuery.prefetch();

  // 3. Dehydrate the cache and pass as page props
  return {
    props: {
      trpcState: helpers.dehydrate(),
    },
  };
};


🚀 The Path Forward: App Router
This exact architectural headache is largely what React Server Components (RSC) and the Next.js App Router were designed to eliminate. With layouts and pages functioning as Server Components by default, hydration caches and complex _app pre-passes are no longer necessary—you can simply await queries directly in the component tree.

For libraries dealing with this in the Pages Router, the server-side helper pattern remains the safest and most standard approach without breaking the Next.js compiler.
